agent/ecscni/plugin_windows.go

```diff
@@ -114,6 +122,46 @@ func (client *cniClient) doSetupNS(ctx context.Context, cfg *Config) (*current.R
 	return curResult, nil
 }
 
+// cleanupNS is called by CleanupNS to cleanup the task namespace by invoking DEL for given CNI configurations.
+func (client *cniClient) cleanupNS(ctx context.Context, cfg *Config) error {
```

Comment:
Other than the guard and the error handling, this function is pretty much the same for Linux and Windows, it doesn't feel right to duplicate it, so I have these questions:

* I don't think this method is invoked in parallel in Linux, I think it's only ever called once to clean up the pause container. Where do you expect parallel invocations for windows? In order to keep the a single method, I think it's safe to add the mutex to linux as well, or just not use it if it's not really needed.
* Given that possibly the only difference for windows would be error handling, we could instead create some `handleDeleteError` function which indicates if the loop should return or keep going. For windows it would keep going, for linux it would return... but ti be honest I think we could do exactly the same on linux.

thoughts?

Reply:
- This method would be evoked in parallel whenever multiple tasks are stopped at once on any instance. The same is true for both Windows and Linux. If we were to add a mutex for Linux as well then it might increase the cleanup time for Linux awsvpc network mode as well. But i think it should be at max 1 minute.

- Yes I agree that we can do the same error handling on Linux as well. Basically, we are trying to delete the networks as much as possible before conceding error. That should be fine for both Linux and Windows.

Let's gather other opinions as well!

Reply:
> * This method would be evoked in parallel whenever multiple tasks are stopped at once on any instance. The same is true for both Windows and Linux. If we were to add a mutex for Linux as well then it might increase the cleanup time for Linux awsvpc network mode as well. But i think it should be at max 1 minute.

I think this is ok since the clean up happens after the task is already in the process of stopping, so there are no strict performance considerations there (plus cleaning up the network is usually really fast). 

> * Yes I agree that we can do the same error handling on Linux as well. Basically, we are trying to delete the networks as much as possible before conceding error. That should be fine for both Linux and Windows.

Yes, I also think this is a good idea for linux anyways

Reply:
regarding guard: is there any reason why windows task networking cleanup needs the lock? my understanding is the cni plugin should be responsible to be thread safe and expect itself to be called in parallel

regarding error handling: yes i agree that it can be added to linux too

Reply:
The CNI plugin calls Windows HNS API's using hcsshim which fails if we call the same API's in parallel.
So if we try to stop multiple tasks at the same time, then multiple CNI plugin invocations would be made at the same time which calls same HNS API. This might lead to failure of few calls.

Therefore, we need to serialize the calls to CNI plugin while cleanup.

Reply:
With respect to the cleanupNS guard, the task workloads for Linux and Windows are different. Windows runs fewer tasks at a time as compared to Linux. If a large number of tasks are stopped at the same time on Linux, then the cleanup time might be greater than expected.

Reply:
As per our offline discussions, the guard has been added only for Windows.
However, for Linux as well, we are performing the cleanup as much as possible before conceding error.